### ISSUE TITLE
[Feature] Trace viewer - more highlights, show assertions

### ISSUE THREAD

**Issue by A-ZC-Lau** [microsoft/playwright#8955]
TLDR
1. There should be more highlights on more actions, i.e. getAttribute
2. Trace viewer should show assertions

It would be nice to have some visual display when getting attributes and making assertions, similarly to Cypress. Currently, the highlighting is limited to certain actions, such as clicking, and there is no information on any assertions (passed or failed).

## Highlighting

1. Lack of highlighting example

Does highlight

`page.locator("button").click()`

Does not highlight

`page.locator("button").getAttribute("title")`

## Assertions

Would be nice to assertions as part of the actions tab or just somewhere to have a better understanding of what and where it exactly failed

# Cypress

With Cypress's trace viewer, shows what element was located, and then what assertion was also made on it

e.g. `cy.get("#username").invoke("attr", "placeholder").should("equal", "please enter your username")`

**Comment by pavelfeldman** [comment]
Is that for your local workflow or for CI/CD? We have all these details about the failure in the test report, you don't even need to open trace for that. But maybe you have a different workflow in mind, could you share more?

**Comment by A-ZC-Lau** [comment]
This is more for local testing, where we are just beginning our testing and may be getting some targeting or something else wrong.

Take for example, cypress

# Cypress

```javascript
it("should find the right a", () => {
	cy.visit("https://getbootstrap.com/");

	cy.get("a").first().should("have.text", "Download");
})
```
when you run the above code, I know exactly why my test failed, because it was getting some element I wasn't expecting (should be getting the second element)

![cypress](https://user-images.githubusercontent.com/20273647/133708831-d779f8ed-4678-4215-943d-140701dcd52a.png)

# Playwright

However, with similar code in playwright, I am not certain why it is failing. (code uses @playwright/test with typescript

```javascript
import { test } from "@playwright/test";

test("should find the right a", async ({ page, context }) => {
	await context.tracing.start({
		name: "test",
		screenshots: true,
		snapshots: true,
	})

	await page.goto("https://getbootstrap.com/");
	
	const text = await page.locator("a").first().innerText();

	await context.tracing.stop({
		path: "trace.zip"
	});
	
	test.expect(text).toBe("Download");
})
```

![playwright](https://user-images.githubusercontent.com/20273647/133709046-86787aa7-f377-4766-85a1-edc17cd42f53.png)

On a somewhat related note, if `test.expect` was placed BEFORE `context.tracing.stop`, and the assertion fails, no trace is actually created.

**Comment by A-ZC-Lau** [comment]
Yes, the report on something that failed is different to knowing why it got those details in the first place.

With the cypress example, you can see why the locator got text that it got, because it was retrieving a hidden/off screen "a" element through the highlight. As opposed to the intended element with the "Download" text.

But if you run it with playwright, you have 0 idea of what element it's getting, because the trace doesn't highlight any elements that you're acting upon, i.e. `.innerText()`

**Comment by pavelfeldman** [comment]
Got it, I think I understand what you mean. We are working on `--debug` that would do pretty much `PWDEBUG=1 npx playwright test`. Trace viewer will also get assertions at some point, but we keep it for the post-mortem story mostly.

**Comment by A-ZC-Lau** [comment]
> Got it, I think I understand what you mean. We are working on `--debug` that would do pretty much `PWDEBUG=1 npx playwright test`, but assertion-oriented. Trace viewer will also get assertions at some point, but we keep it for the post-mortem story mostly.

Sounds great. I think at minimum, showing highlights in the trace when a locator performs any sort of action, e.g.`.innerText()` would be a great minor release, since that functionality already exists for `.click()`.

**Comment by pavelfeldman** [comment]
Some work has been done here, please report on what is missing after 1.16!
